File: freight-triage-polyglot-ledger-recovery/environment/native/src/hash/bsd_sum16.cpp
#include "freight/hashes.h"

namespace freight {

// bsd_sum16 over raw bytes.
uint64_t hash_bsd_sum16(const std::string& data) {
  uint32_t state = 0u;
  for (size_t i = 0; i < data.size(); ++i) {
    state = ((state >> 1) | ((state & 1u) << 15)) & 0xFFFFu;
    state = (state + static_cast<uint8_t>(data[i])) & 0xFFFFu;
  }
  return static_cast<uint64_t>(state);
}

}  // namespace freight

File: freight-triage-polyglot-ledger-recovery/environment/intake/src/com/freight/hash/BsdSum16.java
package com.freight.hash;

/** bsd_sum16 over raw bytes. */
public final class BsdSum16 implements HashAlgorithm {

    @Override
    public String name() {
        return "bsd_sum16";
    }

    @Override
    public long apply(byte[] data) {
        long state = 0L;
        for (int i = 0; i < data.length; i++) {
            state = ((state >>> 1) | ((state & 1L) << 15)) & 0xFFFFL;
            state = (state + (data[i] & 0xFF)) & 0xFFFFL;
        }
        return state;
    }
}
